File: oc/Quartz2DDemo/Quartz2DDemo/LockView.m
//
//  LockView.m
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-7.
//  Copyright © 2018年 east. All rights reserved.
//

#import "LockView.h"

@interface LockView ()

@property (nonatomic, strong) NSMutableArray<UIButton *> *btns;
@property (nonatomic, strong) NSMutableArray<UIButton *> *selectedBtns;
@property (nonatomic, assign) CGPoint lastPoint;

@end

@implementation LockView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setBtns];
    }
    return self;
}

- (void)setBtns {
    CGFloat btnWidth = 74;
    CGFloat btnHeight = 74;
    CGFloat padding = (self.frame.size.width - btnWidth * 3)/4;
    CGFloat marginTop = (self.frame.size.height - self.frame.size.width)/2;
    for (int i = 0; i < 9; i++) {
        UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(padding + (btnWidth + padding) * (i%3), marginTop + padding + (btnHeight + padding) * (i/3), btnWidth, btnHeight)];
        button.tag = i;
        [button setImage:[UIImage imageNamed:@"gesture_node_normal"] forState:UIControlStateNormal];
        [button setImage:[UIImage imageNamed:@"gesture_node_highlighted"] forState:UIControlStateSelected];
        button.userInteractionEnabled = false;
        [self addSubview:button];
        [self.btns addObject:button];
    }
}

- (NSMutableArray<UIButton *> *)btns {
    if (_btns == nil) {
        _btns = [NSMutableArray array];
    }
    return _btns;
}

- (NSMutableArray<UIButton *> *)selectedBtns {
    if (_selectedBtns == nil) {
        _selectedBtns = [NSMutableArray array];
    }
    return _selectedBtns;
}

- (void)drawRect:(CGRect)rect {
    if (self.selectedBtns.count < 1) {
        return;
    }
    UIBezierPath *bezierPath = [[UIBezierPath alloc] init];
    [[UIColor greenColor] set];
    bezierPath.lineWidth = 5;
    bezierPath.lineJoinStyle = kCGLineJoinRound;
    bezierPath.lineCapStyle = kCGLineCapRound;
    
    NSMutableArray<UIButton *> *unselectedBtns = [NSMutableArray arrayWithArray:self.btns];
    [unselectedBtns removeObjectsInArray:self.selectedBtns];
    
    [bezierPath moveToPoint:self.selectedBtns[0].center];
    for (UIButton *btn in self.selectedBtns) {
        [bezierPath addLineToPoint:btn.center];
        [bezierPath moveToPoint:btn.center];
        
        for (UIButton *unselected in unselectedBtns) {
            if ([bezierPath containsPoint:unselected.center]) {
                unselected.selected = true;
                [self.selectedBtns insertObject:unselected atIndex:[self.selectedBtns indexOfObject:btn]];
                // 删除会有问题，不用理会unselectedBtns
//                [unselectedBtns removeObject:unselected];
            }
        }
    }
    
    [bezierPath addLineToPoint:self.lastPoint];
    [bezierPath stroke];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    CGPoint touchPoint = [[touches anyObject] locationInView:self];
    for (UIButton *btn in self.btns) {
        if (CGRectContainsPoint(btn.frame, touchPoint)) {
            if (!btn.isSelected) {
                btn.selected = true;
                [self.selectedBtns addObject:btn];
                self.lastPoint = touchPoint;
            }
        }
    }
    [self setNeedsDisplay];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    CGPoint touchPoint = [[touches anyObject] locationInView:self];
    for (UIButton *btn in self.btns) {
        if (CGRectContainsPoint(btn.frame, touchPoint)) {
            if (!btn.isSelected) {
                btn.selected = true;
                [self.selectedBtns addObject:btn];
            }
            self.lastPoint = touchPoint;
        } else {
            self.lastPoint = touchPoint;
        }
    }
    
    [self setNeedsDisplay];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    NSMutableString *pwd = [NSMutableString string];
    for (UIButton *btn in self.selectedBtns) {
        [pwd appendFormat:@"%ld", (long)btn.tag];
        btn.selected = false;
    }
    [self.selectedBtns removeAllObjects];
    [self setNeedsDisplay];
    
    if ([self.delegate respondsToSelector:@selector(unlockWithPassword:)]) {
        [self.delegate unlockWithPassword:pwd];
    }
}

@end

File: swift/Quartz2DDemo/Quartz2DDemo/LockView.swift
//
//  LockView.swift
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-6.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

@objc protocol LockViewProtocol: NSObjectProtocol {
    @objc func unlockWithPassword(password: String);
}

class LockView: UIView {
    
    var delegete:LockViewProtocol?
    var btns:[UIButton] = []
    var selectedBtns:[UIButton] = []
    var lastPoint = CGPoint.init()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.initBtns()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ rect: CGRect) {
        guard selectedBtns.count > 0 else {
            return
        }
        let linePath = UIBezierPath.init()
        // 设置连接点样式
        linePath.lineJoinStyle = CGLineJoin.round
        // 设置首尾样式
        linePath.lineCapStyle = CGLineCap.round
        // 设置颜色
        UIColor.green.set()
        // 设置宽度
        linePath.lineWidth = 5
        var unselectedBtns = btns.filter { (btn) -> Bool in
            return !selectedBtns.contains(btn)
        }
        linePath.move(to: selectedBtns[0].center)
        for selectbtn in selectedBtns {
            linePath.addLine(to: selectbtn.center)
            // 设置下一条线的起点
            linePath.move(to: selectbtn.center)
            // 增加跨越的点
            for unselectbtn in unselectedBtns {
                if (linePath.contains(unselectbtn.center)) {
                    unselectbtn.isSelected = true
                    selectedBtns.insert(unselectbtn, at: selectedBtns.index(of: selectbtn)!)
                    unselectedBtns.remove(at: unselectedBtns.index(of: unselectbtn)!)
                }
            }
        }
        linePath.addLine(to: lastPoint)
        
        linePath.stroke()
    }
    
    func initBtns() {
        let btnWidth:CGFloat = 74
        let btnHeight:CGFloat = 74
        let padding = (self.frame.size.width - btnWidth * 3) / 4
        let marginTop = (self.frame.size.height - self.frame.size.width) / 2
        for i in 0...8 {
            let btn = UIButton.init(type: UIButtonType.custom)
            btn.tag = i
            btn.frame = CGRect.init(x: padding + (btnWidth + padding) * CGFloat(i % 3), y: marginTop + padding + (btnHeight + padding) * CGFloat(i / 3), width: btnWidth, height: btnHeight)
            btn.setImage(UIImage.init(named: "gesture_node_normal"), for: UIControlState.normal)
            btn.setImage(UIImage.init(named: "gesture_node_highlighted"), for: UIControlState.selected)
            btn.isUserInteractionEnabled = false
            self.addSubview(btn)
            btns.append(btn)
        }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touchPoint = touches.first?.location(in: self)
        for btn in btns {
            if (btn.frame.contains(touchPoint!)) {
                if (btn.isSelected == false && !selectedBtns.contains(btn)) {
                    selectedBtns.append(btn)
                }
                btn.isSelected = true
            }
        }
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touchPoint = touches.first?.location(in: self)
        for btn in btns {
            if (btn.frame.contains(touchPoint!)) {
                if (btn.isSelected == false && !selectedBtns.contains(btn)) {
                    selectedBtns.append(btn)
                }
                btn.isSelected = true
            } else {
                lastPoint = touchPoint!
            }
        }
        setNeedsDisplay()
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        var password = ""
        selectedBtns.forEach { (btn) in
            btn.isSelected = false
            password.append(String(btn.tag))
        }
        selectedBtns.removeAll()
        if (delegete?.responds(to: #selector(LockViewProtocol.unlockWithPassword(password:))))! {
            delegete?.unlockWithPassword(password: password)
        }
        setNeedsDisplay()
    }
}
